toolchain/check/handle_name.cpp

```diff
@@ -168,6 +178,37 @@ auto HandleParseNode(Context& context, Parse::NameQualifierId /*node_id*/)
   return true;
 }
 
+auto HandleParseNode(Context& context, Parse::DesignatorExprId node_id)
+    -> bool {
+  SemIR::NameId name_id = context.node_stack().PopName();
+
+  if (name_id == SemIR::NameId::SelfType) {
+    // Look up `.Self`.
+    SemIR::InstId period_self_id =
+        HandleNameAsExpr(context, node_id, SemIR::NameId::PeriodSelf);
+    context.node_stack().Push(node_id, period_self_id);
+  } else {
+    // Otherwise this is `.Member`, so look up `.Self` and then `Member` in
+    // `.Self`.
+    SemIR::InstId period_self_id = SemIR::InstId::Invalid;
+    {
+      DiagnosticAnnotationScope annotate_diagnostics(
+          &context.emitter(), [&](auto& builder) {
+            CARBON_DIAGNOSTIC(
+                NoPeriodSelfForDesignator, Note,
+                "Designator may only be used when `.Self` is in scope.");
```

Comment:
Minimal style change, considering See #4321:

```suggestion
                "designator may only be used when `.Self` is in scope");
```

But, had you considered changing HandleNameAsExpr to pass `/*required=*/false` to `LookupUnqualifiedName` and emitting as a full diagnostic here, rather than an annotation?

Note, a similar diagnostic might be `CoreNameNotFound` which uses: ``name `Core.{0}` implicitly referenced here, but not found``... So if you had control over the diagnostic, maybe you could just replace the current ``name `.Self` not found`` with a more similar ``name `.Self` implicitly referenced by designated expression, but not found``.

Reply:
Made the style change, and added a TODO to make the more invasive change to the diagnostic.